src/Storages/Kafka/KafkaConsumer2.cpp

```diff
@@ -0,0 +1,375 @@
+        }
+
+        if (new_messages.empty())
+        {
+            LOG_TRACE(log, "Stalled");
+            return nullptr;
+        }
+        else
+        {
+            messages = std::move(new_messages);
+            current = messages.begin();
+            LOG_TRACE(
+                log,
+                "Polled batch of {} messages. Offsets position: {}",
+                messages.size(),
+                consumer->get_offsets_position(consumer->get_assignment()));
+            break;
+        }
+    }
+
+    filterMessageErrors();
+    if (current == messages.end())
+    {
+        LOG_ERROR(log, "Only errors left");
+        stalled_status = StalledStatus::ERRORS_RETURNED;
+        return nullptr;
+    }
+
+    ProfileEvents::increment(ProfileEvents::KafkaMessagesPolled, messages.size());
+
+    stalled_status = StalledStatus::NOT_STALLED;
+    return getNextMessage();
+}
+
+void KafkaConsumer2::commit(const TopicPartition & topic_partition)
+{
+    static constexpr auto max_retries = 5;
+    bool committed = false;
+
+    LOG_TEST(
+        log,
+        "Trying to commit offset {} to Kafka for topic-partition [{}:{}]",
+        topic_partition.offset,
+        topic_partition.topic,
+        topic_partition.partition_id);
+
+    const auto topic_partition_list = std::vector{cppkafka::TopicPartition{
+        topic_partition.topic,
+        topic_partition.partition_id,
+        topic_partition.offset,
+    }};
+    for (auto try_count = 0; try_count < max_retries && !committed; ++try_count)
+    {
+        try
+        {
+            // See https://github.com/edenhill/librdkafka/issues/1470
+            // broker may reject commit if during offsets.commit.timeout.ms (5000 by default),
+            // there were not enough replicas available for the __consumer_offsets topic.
+            // also some other temporary issues like client-server connectivity problems are possible
+
+            consumer->commit(topic_partition_list);
+            committed = true;
+            LOG_INFO(
+                log,
+                "Committed offset {} to Kafka for topic-partition [{}:{}]",
+                topic_partition.offset,
+                topic_partition.topic,
+                topic_partition.partition_id);
+        }
+        catch (const cppkafka::HandleException & e)
+        {
+            // If there were actually no offsets to commit, return. Retrying won't solve
+            // anything here
+            if (e.get_error() == RD_KAFKA_RESP_ERR__NO_OFFSET)
+                committed = true;
+            else
+                LOG_WARNING(log, "Exception during commit attempt: {}", e.what());
+        }
+    }
+
+    if (!committed)
+    {
+        // The failure is not the biggest issue, it only counts when a table is dropped and recreated, otherwise the offsets are taken from keeper.
+        ProfileEvents::increment(ProfileEvents::KafkaCommitFailures);
+        LOG_INFO(log, "All commit attempts failed");
```

Comment:
> How normal is this?

This shouldn't really happen, I am honestly don't see scenario when this would happen without the whole Kafka cluster is being down. In the old Kafka engine, we throw an exception. I wouldn't like to throw an exception here, as we don't really rely on the offsets in Kafka once the table is created.

> I am afraid here commit to kafka might get broken accidentally in the future and we will not notice at all, because it is just LOG_INFO.

That is a valid point. I raised the log level to error to make it more visible. however we have a test that specifically relies on committing  to and reading the offset from Kafka in case of table recreation.

> what is the worst case scenario that we fail to commit to kafka while everything else is successful

If the table is not recreated, then nothing. If the table is recreated, then we might read messages twice.

> Will the messages be always stuck in kafka forever or they will be automatically commited with the next offset commit?

In Kafka only one offset is stored for each topic-partition. Basically if committing offset `x` fails, but with the next batch we can successfully commit offset `y` (where `y > x`), then failing to commit `x` has no affect on anything.

> Should we also retry the commit to kafka if error is retriable?

At this point we tried to commit 5 times already, I don't think it makes sense to try any longer.